[CI] Stabilize durable command-WAL standalone visibility on Windows

Parent CI tracker: #3776

Discovered while proving: #3796 / PR #3798

Related but distinct completed ownership fix: #3778 / PR #3825

## Goal

Make the standalone durable command-WAL BSON visibility regression a stable Windows CI contract: preserve meaningful durable insert/read/update/read coverage, identify whether the current failure is an implementation latency regression or an oversized smoke workload with a shared phase deadline, and remove the hosted 60-second cliff without hiding a stuck request or weakening durability.

## Exact failure

Exact candidate `e7a172a8497d536e9d12953b97764f9abd7f43b4` failed independent proof-A TreeDB run [29573559168](https://github.com/snissn/gomap/actions/runs/29573559168), attempt 2, job [`windows-core-3`](https://github.com/snissn/gomap/actions/runs/29573559168/job/87867002064):

```text
--- PASS: TestStandaloneServerCommandWALBSONInsertReadUpdateVisibility/command_wal_relaxed (1.17s)
--- FAIL: TestStandaloneServerCommandWALBSONInsertReadUpdateVisibility/command_wal_durable (60.16s)
    standalone_test.go:796: update user1997: ... context deadline exceeded ...
```

The candidate changes one caching waiter assertion and is disjoint from `TreeDB/mongo_gateway`. All other 14 TreeDB jobs passed. The failure was not rerun.

The test currently creates one operation context for the complete 2,048-document scenario. That same 60-second context covers ping, 16-worker inserts, all reads, and 2,048 sequential update-plus-read pairs. Expiration at `user1997` strongly suggests cumulative durable-path wall time reached the shared deadline near the end of the loop; it does not by itself show that the final update was stuck for 60 seconds.

This is distinct from #3778, which fixed TreeDB-owned Windows file handles blocking covered command-WAL segment removal.

## Test-first characterization

Before changing product code, workload size, or deadlines:

1. Add or use bounded phase/operation timing evidence that distinguishes cumulative phase-budget exhaustion from an individual blocked request.
2. Reproduce the durable profile repeatedly on the exact focused test locally where useful and on hosted `windows-core-3`; record insert, read, update, and final-read progress/timing without turning CI output into an unbounded performance log.
3. If product latency or blocking is responsible, add the smallest deterministic regression for that mechanism and fix it without weakening command-WAL sync, visibility, recovery, or stable resource ownership.
4. If the test is functioning as an accidental 2,048-sync performance gate, replace the cliff with a bounded smoke contract that still exercises concurrent inserts, sequential durable updates, BSON projection, immediate visibility, clean shutdown, and both relaxed/durable profiles. Any workload reduction must be justified by retained semantic coverage and accompanied by a separate guard against a single hung request.

## Boundaries

- Do not simply increase the 60-second timeout or add retries/sleeps.
- Do not disable durable command-WAL sync or make the durable profile use relaxed semantics.
- Do not drop update/read-after-update visibility assertions.
- Do not treat a client timeout as proof of corruption or a server deadlock without evidence.
- Do not reopen or conflate #3778 unless the exact released-resource ownership failure recurs.
- Preserve Windows coverage; a Unix-only green result is insufficient.

## Performance classification

Performance-relevant until characterized. Report the exact document count, per-phase wall times, platform, SHA, and durable/relaxed profile. A product change requires before/after focused evidence. A test-only scale or deadline change must explain why the old value was not a calibrated performance threshold and how per-request hang detection remains bounded.

## Required evidence

- [ ] A deterministic or repeated characterization identifies cumulative phase exhaustion versus an individually stuck operation.
- [ ] Focused normal runs pass repeatedly for both profiles.
- [ ] Focused race coverage passes for the affected package/path.
- [ ] Hosted exact-head `windows-core-3` passes repeatedly without relying on a failed-job rerun as primary evidence.
- [ ] The full exact-head TreeDB matrix plus two independent exact-SHA proof matrices are green.
- [ ] Exact-head Codex review is clean and all review threads are resolved.
- [ ] Exact default-branch TreeDB CI is green after merge.

## Graph role

This node precedes final proof of #3796 / PR #3798 and final certification #3777. It is independent of #3843's text-index catalog-ordering fix and may be implemented separately, but hosted TreeDB matrices remain serialized under the tracker ceiling.

--- a/.github/scripts/test_treedb_windows_core_headroom.py
+++ b/.github/scripts/test_treedb_windows_core_headroom.py
@@ -132,6 +132,34 @@ class TreeDBWindowsCoreHeadroomTest(unittest.TestCase):
                 self.assertEqual(matrix_timeout(workflow, job_name), 25)
 
 
+class TreeDBLinuxAllPackagesHeadroomTest(unittest.TestCase):
+    def test_ubuntu_keeps_complete_serial_coverage_inside_a_bounded_cap(self) -> None:
+        workflow = WORKFLOW.read_text(encoding="utf-8")
+        matrix_match = re.search(
+            r"- name: ubuntu-latest\s+"
+            r"os: ubuntu-latest\s+"
+            r"timeout: (?P<timeout>\d+)\s+"
+            r"shard_kind: (?P<shard_kind>\S+)\s+"
+            r"run_vet: (?P<run_vet>\S+)",
+            workflow,
+        )
+        self.assertIsNotNone(matrix_match, "missing Ubuntu all-package matrix entry")
+
+        # The failed hosted run consumed about 14m13s in vet plus the four
+        # dominant packages before the fixed 15-minute cap canceled it. Keep
+        # a bounded cap with meaningful runner-variability and artifact margin.
+        self.assertEqual(int(matrix_match.group("timeout")), 25)
+        self.assertEqual(matrix_match.group("shard_kind"), "all")
+        self.assertEqual(matrix_match.group("run_vet"), "true")
+
+        test_job = workflow_job(workflow, "test")
+        self.assertIn("GOMAXPROCS: 2", test_job)
+        self.assertIn(
+            "go test -json -timeout 30m -p 1 ./... | tee treedb-test.jsonl",
+            test_job,
+        )
+
+
 class TreeDBLinuxRaceHeadroomTest(unittest.TestCase):
     def test_db_package_timeout_is_bounded_inside_the_job_cap(self) -> None:
         workflow = WORKFLOW.read_text(encoding="utf-8")

--- a/.github/workflows/treedb-tests.yml
+++ b/.github/workflows/treedb-tests.yml
@@ -178,7 +178,7 @@ jobs:
         include:
           - name: ubuntu-latest
             os: ubuntu-latest
-            timeout: 15
+            timeout: 25
             shard_kind: all
             run_vet: true
           - name: macos-latest

--- a/TreeDB/mongo_gateway/standalone_test.go
+++ b/TreeDB/mongo_gateway/standalone_test.go
@@ -661,22 +661,23 @@ func TestStandaloneServerCommandWALBSONInsertReadUpdateVisibility(t *testing.T) 
 				t.Fatalf("mongo connect: %v", err)
 			}
 
-			opTimeout := 15 * time.Second
-			if profile == treedb.ProfileCommandWALDurable {
-				opTimeout = 60 * time.Second
-			}
-			opCtx, opCancel := context.WithTimeout(context.Background(), opTimeout)
-			defer opCancel()
-			if err := client.Ping(opCtx, nil); err != nil {
+			const requestTimeout = 10 * time.Second
+			scenarioStarted := time.Now()
+			pingStarted := time.Now()
+			pingCtx, pingCancel := context.WithTimeout(context.Background(), requestTimeout)
+			err = client.Ping(pingCtx, nil)
+			pingElapsed := time.Since(pingStarted)
+			pingCancel()
+			if err != nil {
 				_ = client.Disconnect(context.Background())
 				cancel()
 				_ = ln.Close()
 				_ = standalone.Close()
-				t.Fatalf("driver ping: %v", err)
+				t.Fatalf("driver ping after %s (request limit %s): %v", pingElapsed, requestTimeout, err)
 			}
 
 			coll := client.Database("ycsb").Collection("usertable")
-			const docs = 2048
+			const docs = 512
 			const workers = 16
 			ycsbValues := func(i int) bson.D {
 				values := make(bson.D, 0, 11)
@@ -699,36 +700,45 @@ func TestStandaloneServerCommandWALBSONInsertReadUpdateVisibility(t *testing.T) 
 			}
 
 			type insertError struct {
-				index int
-				err   error
+				index   int
+				elapsed time.Duration
+				err     error
 			}
 			errCh := make(chan insertError, 1)
+			insertTimingCh := make(chan time.Duration, docs)
 			workCh := make(chan int)
+			insertCtx, insertCancel := context.WithCancel(context.Background())
 			var insertWG sync.WaitGroup
-			sendInsertErr := func(i int, err error) {
+			sendInsertErr := func(i int, elapsed time.Duration, err error) {
 				if err == nil {
 					return
 				}
 				select {
-				case errCh <- insertError{index: i, err: err}:
+				case errCh <- insertError{index: i, elapsed: elapsed, err: err}:
 				default:
 				}
-				opCancel()
+				insertCancel()
 			}
+			insertStarted := time.Now()
 			for worker := 0; worker < workers; worker++ {
 				insertWG.Add(1)
 				go func() {
 					defer insertWG.Done()
 					for {
 						select {
-						case <-opCtx.Done():
+						case <-insertCtx.Done():
 							return
 						case i, ok := <-workCh:
 							if !ok {
 								return
 							}
-							_, err := coll.InsertOne(opCtx, ycsbValues(i))
-							sendInsertErr(i, err)
+							requestStarted := time.Now()
+							requestCtx, requestCancel := context.WithTimeout(insertCtx, requestTimeout)
+							_, err := coll.InsertOne(requestCtx, ycsbValues(i))
+							requestElapsed := time.Since(requestStarted)
+							requestCancel()
+							insertTimingCh <- requestElapsed
+							sendInsertErr(i, requestElapsed, err)
 						}
 					}
 				}()
@@ -737,7 +747,7 @@ func TestStandaloneServerCommandWALBSONInsertReadUpdateVisibility(t *testing.T) 
 				defer close(workCh)
 				for i := 0; i < docs; i++ {
 					select {
-					case <-opCtx.Done():
+					case <-insertCtx.Done():
 						return
 					case workCh <- i:
 					}
@@ -746,15 +756,18 @@ func TestStandaloneServerCommandWALBSONInsertReadUpdateVisibility(t *testing.T) 
 			insertDone := make(chan struct{})
 			go func() {
 				insertWG.Wait()
+				close(insertTimingCh)
 				close(insertDone)
 			}()
 			select {
 			case insertErr := <-errCh:
+				insertCancel()
+				<-insertDone
 				_ = client.Disconnect(context.Background())
 				cancel()
 				_ = ln.Close()
 				_ = standalone.Close()
-				t.Fatalf("insert %d: %v", insertErr.index, insertErr.err)
+				t.Fatalf("insert %d after %s (request limit %s): %v", insertErr.index, insertErr.elapsed, requestTimeout, insertErr.err)
 			case <-insertDone:
 				select {
 				case insertErr := <-errCh:
@@ -762,38 +775,62 @@ func TestStandaloneServerCommandWALBSONInsertReadUpdateVisibility(t *testing.T) 
 					cancel()
 					_ = ln.Close()
 					_ = standalone.Close()
-					t.Fatalf("insert %d: %v", insertErr.index, insertErr.err)
+					t.Fatalf("insert %d after %s (request limit %s): %v", insertErr.index, insertErr.elapsed, requestTimeout, insertErr.err)
 				default:
 				}
-				if err := opCtx.Err(); err != nil {
-					_ = client.Disconnect(context.Background())
-					cancel()
-					_ = ln.Close()
-					_ = standalone.Close()
-					t.Fatalf("insert workers stopped: %v", err)
+			}
+			insertCancel()
+			insertElapsed := time.Since(insertStarted)
+			var insertMax time.Duration
+			for elapsed := range insertTimingCh {
+				if elapsed > insertMax {
+					insertMax = elapsed
 				}
 			}
 
+			var initialReadElapsed time.Duration
+			var initialReadMax time.Duration
+			var updateElapsed time.Duration
+			var updateMax time.Duration
+			var finalReadElapsed time.Duration
+			var finalReadMax time.Duration
 			for i := 0; i < docs; i++ {
 				id := "user" + strconv.Itoa(i)
 				var got map[string][]byte
-				if err := coll.FindOne(opCtx, bson.D{{Key: "_id", Value: id}}, options.FindOne().SetProjection(projection)).Decode(&got); err != nil {
-					_ = client.Disconnect(context.Background())
-					cancel()
-					_ = ln.Close()
-					_ = standalone.Close()
-					t.Fatalf("find after insert %s: %v", id, err)
+				requestStarted := time.Now()
+				requestCtx, requestCancel := context.WithTimeout(context.Background(), requestTimeout)
+				err := coll.FindOne(requestCtx, bson.D{{Key: "_id", Value: id}}, options.FindOne().SetProjection(projection)).Decode(&got)
+				requestElapsed := time.Since(requestStarted)
+				requestCancel()
+				initialReadElapsed += requestElapsed
+				if requestElapsed > initialReadMax {
+					initialReadMax = requestElapsed
 				}
-				result, err := coll.UpdateOne(opCtx,
-					bson.D{{Key: "_id", Value: id}},
-					bson.D{{Key: "$set", Value: ycsbUpdate(i)}},
-				)
 				if err != nil {
 					_ = client.Disconnect(context.Background())
 					cancel()
 					_ = ln.Close()
 					_ = standalone.Close()
-					t.Fatalf("update %s: %v", id, err)
+					t.Fatalf("find after insert %s after %s (request limit %s): %v", id, requestElapsed, requestTimeout, err)
+				}
+				requestStarted = time.Now()
+				requestCtx, requestCancel = context.WithTimeout(context.Background(), requestTimeout)
+				result, err := coll.UpdateOne(requestCtx,
+					bson.D{{Key: "_id", Value: id}},
+					bson.D{{Key: "$set", Value: ycsbUpdate(i)}},
+				)
+				requestElapsed = time.Since(requestStarted)
+				requestCancel()
+				updateElapsed += requestElapsed
+				if requestElapsed > updateMax {
+					updateMax = requestElapsed
+				}
+				if err != nil {
+					_ = client.Disconnect(context.Background())
+					cancel()
+					_ = ln.Close()
+					_ = standalone.Close()
+					t.Fatalf("update %s after %s (request limit %s): %v", id, requestElapsed, requestTimeout, err)
 				}
 				if result.MatchedCount != 1 || result.ModifiedCount != 1 {
 					_ = client.Disconnect(context.Background())
@@ -803,12 +840,21 @@ func TestStandaloneServerCommandWALBSONInsertReadUpdateVisibility(t *testing.T) 
 					t.Fatalf("update %s matched=%d modified=%d, want 1/1", id, result.MatchedCount, result.ModifiedCount)
 				}
 				got = nil
-				if err := coll.FindOne(opCtx, bson.D{{Key: "_id", Value: id}}, options.FindOne().SetProjection(projection)).Decode(&got); err != nil {
+				requestStarted = time.Now()
+				requestCtx, requestCancel = context.WithTimeout(context.Background(), requestTimeout)
+				err = coll.FindOne(requestCtx, bson.D{{Key: "_id", Value: id}}, options.FindOne().SetProjection(projection)).Decode(&got)
+				requestElapsed = time.Since(requestStarted)
+				requestCancel()
+				finalReadElapsed += requestElapsed
+				if requestElapsed > finalReadMax {
+					finalReadMax = requestElapsed
+				}
+				if err != nil {
 					_ = client.Disconnect(context.Background())
 					cancel()
 					_ = ln.Close()
 					_ = standalone.Close()
-					t.Fatalf("find after update %s: %v", id, err)
+					t.Fatalf("find after update %s after %s (request limit %s): %v", id, requestElapsed, requestTimeout, err)
 				}
 				if string(got["field0"]) != "updated"+strconv.Itoa(i)+"-0" {
 					_ = client.Disconnect(context.Background())
@@ -818,6 +864,8 @@ func TestStandaloneServerCommandWALBSONInsertReadUpdateVisibility(t *testing.T) 
 					t.Fatalf("find after update %s field0=%q want updated", id, got["field0"])
 				}
 			}
+			t.Logf("command-WAL BSON visibility profile=%s docs=%d workers=%d request_limit=%s ping=%s insert_wall=%s insert_max=%s initial_read_total=%s initial_read_max=%s update_total=%s update_max=%s final_read_total=%s final_read_max=%s scenario_wall=%s",
+				profile, docs, workers, requestTimeout, pingElapsed, insertElapsed, insertMax, initialReadElapsed, initialReadMax, updateElapsed, updateMax, finalReadElapsed, finalReadMax, time.Since(scenarioStarted))
 
 			if err := client.Disconnect(context.Background()); err != nil {
 				cancel()

